btc-withdrawal/src/main/kotlin/withdrawal/btc/transaction/TransactionHelper.kt

```diff
@@ -126,15 +105,97 @@ class TransactionHelper(
+     * @param amount - amount of SAT to spend
+     * @param confidenceLevel - minimum depth of transactions
+     * @param recursivelyCollectedUnspents - list of unspents collected from all recursion levels. It will be returned at the end on execution
+     * @return list full of unspent transactions
+     */
+    private tailrec fun collectUnspentsRec(
+        availableAddresses: Set<String>,
+        wallet: Wallet,
+        amount: Long,
+        confidenceLevel: Int,
+        recursivelyCollectedUnspents: MutableList<TransactionOutput>
+    ): List<TransactionOutput> {
+        //Only confirmed transactions must be used
+        val unspents =
+            ArrayList<TransactionOutput>(wallet.unspents.filter { unspent ->
+                unspent.parentTransactionDepthInBlocks >= confidenceLevel
+                        //Cannot use already used unspents
+                        && !usedOutputs.contains(unspent)
+                        //Cannot use unspents from another level of recursion
+                        && !recursivelyCollectedUnspents.contains(unspent)
+            })
+        if (unspents.isEmpty()) {
+            throw IllegalStateException("Out of unspents")
+        }
+        /*
+        Wallet stores unspents in a HashSet. Order of a HashSet depends on several factors: current array size and etc.
+        This may lead different notary nodes to pick different transactions.
+        This is why we order transactions manually, essentially reducing the probability of
+        different nodes to pick different output transactions.*/
+        unspents.sortWith(Comparator { output1, output2 ->
+            /*
+            Outputs are compared by values.
+            It will help us having a little amount of inputs.
+            Less inputs -> smaller tx size -> smaller fee*/
+            val valueComparison = output1.value.value.compareTo(output2.value.value)
+            //If values are the same, we compare by hash codes
+            if (valueComparison == 0) {
+                output1.hashCode().compareTo(output2.hashCode())
+            } else {
+                valueComparison
+            }
+        })
+        var collectedAmount = getTotalUnspentValue(recursivelyCollectedUnspents)
+        //We use registered clients outputs only
+        unspents.filter { unspent -> isAvailableOutput(availableAddresses, unspent) }.forEach { unspent ->
+            if (collectedAmount >= amount) {
+                return@forEach
+            }
+            collectedAmount += unspent.value.value
+            recursivelyCollectedUnspents.add(unspent)
+        }
+        if (collectedAmount < amount) {
+            throw IllegalStateException("Cannot get enough BTC amount(required $amount, collected $collectedAmount) using current unspent tx collection")
+        }
+        //Check if able to pay fee
+        val amountAndFee = amount + getTxFee(recursivelyCollectedUnspents.size, OUTPUTS)
```

Comment:
But this will add recursively fee each time, right?

Reply:
Better use const amount and dynamic fee variable, otherwise this recursion will not terminate optimally